url1a745d67fc_sobercheg_InterviewTraining_tgz-pJ8-leetcode_oj_Solution2TestJ8/src/leetcode/oj/Solution.java

```diff
@@ -0,0 +1,3742 @@
+     * Find all unique quadruplets in the array which gives the sum of target.
+     * <p/>
+     * Note:
+     * <p/>
+     * Elements in a quadruplet (a,b,c,d) must be in non-descending order. (ie, a ≤ b ≤ c ≤ d)
+     * The solution set must not contain duplicate quadruplets.
+     * <p/>
+     * For example, given array S = {1 0 -1 0 -2 2}, and target = 0.
+     * <p/>
+     * A solution set is:
+     * (-1,  0, 0, 1)
+     * (-2, -1, 1, 2)
+     * (-2,  0, 0, 2)
+     */
+    public List<List<Integer>> fourSum(int[] num, int target) {
+        if (num.length < 4) return new ArrayList<List<Integer>>();
+        List<List<Integer>> kSums = kSum(num, target, 4);
+        return new ArrayList<List<Integer>>(kSums);
+    }
+
+    /**
+     * <a href="http://cs.stackexchange.com/questions/2973/generalised-3sum-k-sum-problem">k-SUM problem</a>
+     * <p/>
+     * For even k: Compute a sorted list S of all sums of k/2 input elements.
+     * Check whether S contains both some number x and its negation −x. The algorithm runs in O(nk/2logn) time.
+     * <p/>
+     * For odd k: Compute the sorted list S of all sums of (k−1)/2 input elements.
+     * For each input element a, check whether S contains both x and a−x, for some number x.
+     * (The second step is essentially the O(n2)-time algorithm for 3SUM.) The algorithm runs in O(n(k+1)/2) time.
+     */
+    public List<List<Integer>> kSum(int[] num, int target, int k) {
+        if (num.length < k) return new ArrayList<List<Integer>>();
+        // Step 1. Compute all sums of k/2 input elements
+        // Probably this step brings most complexity and running time
+        // Probably the algo can be redesigned to use n choose k rather than n choose k/2. It'll greatly simplify
+        // the code avoiding combination steps
+        List<List<Integer>> nChooseKIndices = choose(num.length, k / 2);
+        Map<Integer, List<List<Integer>>> sumToCombinations = new HashMap<Integer, List<List<Integer>>>();
+        for (List<Integer> indices : nChooseKIndices) {
+            int sum = 0;
+            for (int index : indices) {
+                sum += num[index];
+            }
+            if (!sumToCombinations.containsKey(sum)) {
+                sumToCombinations.put(sum, new ArrayList<List<Integer>>());
+            }
+            sumToCombinations.get(sum).add(indices);
+        }
+
+        Set<List<Integer>> kSums = new HashSet<List<Integer>>();
+
+        // Step 2. Check whether S contains both some number x and its sibling target−x
+        if (k % 2 == 0) {
+            for (int sum : sumToCombinations.keySet()) {
+                int antiSum = target - sum;
+                if (sumToCombinations.containsKey(antiSum)) {
+                    List<List<Integer>> finalCombos = new ArrayList<List<Integer>>();
+                    for (int i = 0; i < sumToCombinations.get(sum).size(); i++) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.